src/domain/gpu/Solver.ts

```diff
@@ -0,0 +1,147 @@
+                        KOCHAĆ JAKOBINA!!!!!!!!!!!!!!!!!!                                                                                           
+ */
+
+export default class SolverGPU {
+  private gpu: GPU;
+
+  constructor() {
+    this.gpu = new GPU();
+  }
+
+  solve(dataIn: DataInMessagePayload): DataOutMessagePayload {
+    const dataOut: DataOutMessagePayload = {
+      matches: null,
+      error: null,
+    };
+
+    try {
+      this.checkInputPayload(dataIn);
+      dataOut.matches = this._solve(dataIn);
+    } catch (error) {
+      dataOut.error = error;
+    }
+
+    return dataOut;
+  }
+
+  private checkInputPayload(data: DataInMessagePayload) {
+    if (data.text.length < 1) {
+      throw Error(
+        `Text must be at least 1 character long (is ${data.text.length}).`
+      );
+    }
+    if (data.pattern.length < 1) {
+      throw Error(
+        `Pattern must be at least 1 character long (is ${data.pattern.length}).`
+      );
+    }
+    if (data.pattern.length > data.text.length) {
+      throw Error(
+        `Pattern can't be longer than text 
+            (pattern length: ${data.pattern.length}, 
+            text length: ${data.text.length})`
+      );
+    }
+  }
+
+  private _solve(data: DataInMessagePayload): Array<number> {
+    const maxValidMatchPosition = data.text.length - data.pattern.length;
+
+    // Text length rounded up to nearest 128 multiple + text end overlap
+    const paddedTextLen =
+      Math.ceil(data.text.length / 128) * 128 + data.pattern.length - 1;
+    const paddedText = new Uint8Array(paddedTextLen);
+    paddedText.set(data.text, 0);
+
+    const kernelCount = Math.ceil(data.text.length / 128);
+
+    console.log("\nKernelCount=", kernelCount);
+    console.log("\npaddedTextLen=", paddedTextLen);
+
+    const searchKernel = this.gpu
+      .createKernel(this.kernelFunction)
+      .setOutput([kernelCount]);
+
+    const output = searchKernel(
+      (paddedText as unknown) as number[],
+      (data.pattern as unknown) as number[],
+      data.pattern.length
+    );
+
+    const matches: number[] = [];
+
+    (output as Float32Array[]).forEach((chunkArray, chunkArrayIdx) => {
+      // chunkArray contains 4, 32-bit numbers
+      chunkArray.forEach((chunk, chunkIdx) => {
+        // Chunk is a 32-bit number
+        for (let bitIdx = 0; bitIdx < 32; bitIdx++) {
+          if (chunk & (1 << bitIdx)) {
+            // Found match
+            matches.push(chunkArrayIdx * 128 + chunkIdx * 32 + bitIdx);
+            // bitPerPos.push(1);
+          } else {
+            // bitPerPos.push(0);
+          }
+        }
+      });
+    });
+
+    // Filter out matches at pos>maxMatchPosition
+    const filteredMatches: number[] = [];
+    matches.forEach((pos) =>
+      pos <= maxValidMatchPosition ? filteredMatches.push(pos) : null
+    );
+
+    return matches;
+  }
+
+  /**
+   * Each kernel handles 128 positions.
+   * Returns Array<number> of size 4.
```

Comment:
Is it worth to have this configurable?
In CPU we can set number of workers `[1; inf]`.
So here we can set size of data processed by single kernel `[1; 4]`.

Reply:
I suggest you/we test this for variety of cases since I'm not sure about 32b int representation.
Does `console.log` log instance of `Uint32Array`?

Reply:
We could, but there is already quite a bit of information we will get from the GPU vs CPU implementations. Benchmarking and analysis of that many cases will be time-consuming.

Reply:
What do you think @plonajakub ?

Reply:
Regarding data types, maybe we should look into the following:
```typescript
kernel.setPrecision('unsigned' | 'single');
```

Reply:
Yeah, it can be only seen in the performance tests if there is a need of having this configurable. I would discuss it face to face.

Reply:
We decided with @plonajakub to make kernel coverage configurable to 1,2,4 32-bit words (32,64,128 exact matching positions). Should be done by November 4th.